packages/snaps-controllers/src/snaps/location/npm.test.ts

```diff
@@ -0,0 +1,149 @@
+import { assert } from '@metamask/utils';
+import { createReadStream } from 'fs';
+import { readFile } from 'fs/promises';
+import fetchMock from 'jest-fetch-mock';
+import path from 'path';
+
+import { NpmLocation } from './npm';
+
+describe('NpmLocation', () => {
+  beforeEach(() => {
+    fetchMock.resetMocks();
+  });
+
+  it('fetches a package tarball, extracts the necessary files, and validates them', async () => {
+    const { version: templateSnapVersion } = JSON.parse(
+      (
+        await readFile(require.resolve('@metamask/template-snap/package.json'))
+      ).toString('utf8'),
+    );
+
+    const tarballUrl = `https://registry.npmjs.cf/@metamask/template-snap/-/template-snap-${templateSnapVersion}.tgz`;
+    const tarballRegistry = `https://registry.npmjs.org/@metamask/template-snap/-/template-snap-${templateSnapVersion}.tgz`;
+    fetchMock
+      .mockResponseOnce(
+        JSON.stringify({
+          // eslint-disable-next-line @typescript-eslint/naming-convention
+          'dist-tags': {
+            latest: templateSnapVersion,
+          },
+          versions: {
+            [templateSnapVersion]: {
+              dist: {
+                // return npmjs.org registry here so that we can check overriding it with npmjs.cf works
+                tarball: tarballRegistry,
+              },
+            },
+          },
+        }),
+      )
+      .mockResponseOnce(
+        (_req) =>
+          Promise.resolve({
+            ok: true,
+            body: createReadStream(
+              path.resolve(
+                __dirname,
+                `../../../test/fixtures/metamask-template-snap-${templateSnapVersion}.tgz`,
+              ),
+            ),
+          }) as any,
```

Comment:
This was copied from previous npm fetching tests.

Fetch mock allows returning streams instead of simple strings
<img width="531" alt="image" src="https://user-images.githubusercontent.com/1614945/205303576-d61e390e-a47e-443a-aa78-b690ce3cc9d8.png">